libnmstate/dns.py

```diff
@@ -0,0 +1,205 @@
+        if des_dns_state:
+            self._dns_state = des_dns_state
+            if self._config_servers != _get_config_servers(
+                cur_dns_state
+            ) or self._config_searches != _get_config_searches(cur_dns_state):
+                self._config_changed = True
+            self._validate_dns()
+        else:
+            self._dns_state = cur_dns_state if cur_dns_state else {}
+        if ifaces and routes:
+            self._gen_metadata()
+
+    @property
+    def current_config(self):
+        return _get_config(self._cur_dns_state)
+
+    @property
+    def config(self):
+        return _get_config(self._dns_state)
+
+    @property
+    def _config_servers(self):
+        return _get_config_servers(self._dns_state)
+
+    @property
+    def _config_searches(self):
+        return _get_config_searches(self._dns_state)
+
+    def _gen_metadata(self):
+        """
+        Store DNS configurations to interface as metadata
+        """
+        if not self._config_servers and not self._config_searches:
+            return
+        ipv4_iface, ipv6_iface = self._find_ifaces_for_name_servers()
+        index = 0
+        searches_saved = False
+        for server in self._config_servers:
+            iface_name = None
+            if is_ipv6_address(server):
+                iface_name = ipv6_iface
+                family = Interface.IPV6
+            else:
+                iface_name = ipv4_iface
+                family = Interface.IPV4
+            if not iface_name:
+                raise NmstateValueError(
+                    "Failed to find suitable interface for saving DNS "
+                    "name servers: %s" % server
+                )
+            iface = self._ifaces[iface_name]
+            if self._config_changed:
+                iface.mark_as_changed()
+
+            if DnsState.DNS_METADATA not in iface.raw[family]:
+                iface.raw[family][DnsState.DNS_METADATA] = {
+                    DNS.SERVER: [server],
+                    DNS.SEARCH: []
+                    if searches_saved
+                    else self._config_searches,
+                    DnsState.PRIORITY_METADATA: index,
+                }
+            else:
+                iface.raw[family][DnsState.DNS_METADATA][DNS.SERVER].append(
+                    server
+                )
+            searches_saved = True
+            index += 1
+
+    def _find_ifaces_for_name_servers(self):
+        """
+        Find interface to store the DNS configurations in the order of:
+            * Any interface with static gateway
+            * Any interface configured as dynamic IP with 'auto-dns:False'
+        Return tuple: (ipv4_iface, ipv6_iface)
+        """
+        ipv4_iface, ipv6_iface = self._find_ifaces_with_static_gateways()
+        if not (ipv4_iface and ipv6_iface):
+            (
+                new_ipv4_iface,
+                new_ipv6_iface,
```

Comment:
Why do you call these new* ? Aren't these dynamic* ?

Reply:
Changed to `auto_ipv4_iface` and `auto_ipv6_iface`